# Hands-On Exercise 1: E-Commerce Customer Analysis

## Introduction

This exercise focuses on using Pandas to load, inspect, clean, transform, and analyze an e-commerce customer transaction dataset. The analysis will also involve identifying data-quality issues and using the cleaned data to answer basic business questions.

In [1]:
import pandas as pd ## --- this line of code imports the library and prepares it for analysis

In [5]:
df = pd.read_csv("customer_sales.csv") ## --- this line of code imports the dataset for analysis

## Dataset Inspection

The first five records of the dataset will be displayed to obtain an initial understanding of the structure and contents of the data. This helps identify the variables available and provides an early indication of whether the dataset has loaded correctly.

In [6]:
df.head() 

,Customer ID,Customer Name,City,Age,Product Category,Sales,Order Status
0,CUST00001,Joseph Obi,Lagos,65,Beauty,61241.84,Completed
1,CUST00002,Tolu Onyeka,Abuja,55,Home Appliances,7969.88,Completed
2,CUST00003,Fatima Ogunleye,Ilorin,19,Accessories,49729.44,Pending
3,CUST00004,Mary Akinyemi,Ibadan,46,Office Supplies,69565.72,Returned
4,CUST00005,Kelechi Abubakar,Owerri,45,Sports,69485.89,Completed


### Interpretation

The first five records provide an initial view of the structure and contents of the e-commerce customer dataset. The dataset contains information on customer identification, customer name, city, age, product category, sales value, and order status. The displayed records also show that customers are located in different cities and that sales values and order statuses vary across records. This initial inspection confirms that the dataset has been successfully loaded into the Pandas DataFrame and provides a basis for further investigation of its structure and data quality.

In [8]:
df.shape

(5000, 7)

## Interpretation 
the "df.shape" function shows clearly that the dataset contains 5000 records and 7 variables

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Customer ID       5000 non-null   object 
 1   Customer Name     5000 non-null   object 
 2   City              5000 non-null   object 
 3   Age               5000 non-null   int64  
 4   Product Category  5000 non-null   object 
 5   Sales             5000 non-null   float64
 6   Order Status      5000 non-null   object 
dtypes: float64(1), int64(1), object(5)
memory usage: 273.6+ KB


### Interpretation

The `df.info()` output confirms that the dataset contains 5,000 entries and 7 columns. The variables include five columns stored as `object` data types and two numerical columns. The `Age` variable is stored as an integer (`int64`), while `Sales` is stored as a floating-point number (`float64`), which is appropriate because the sales values contain decimal figures. All seven columns contain 5,000 non-null values, indicating that no missing values are currently identified from this initial structural inspection.

In [10]:
df.describe()

,Age,Sales
count,5000.000000,5000.000000
mean,43.905400,124393.017430
std,15.358881,71795.872349
min,18.000000,37.880000
25%,31.000000,61591.572500
50%,44.000000,124713.010000
75%,57.000000,186804.632500
max,70.000000,249810.490000


### Interpretation

The descriptive statistics show that the dataset contains 5,000 valid observations for both age and sales. The average age of the customers is approximately 43.91 years, with ages ranging from 18 to 70 years. The mean sales value is approximately 124,393.02, while the recorded sales values range from 37.88 to 249,810.49. The median age is 44 years, while the median sales value is 124,713.01.

In [11]:
df.isnull().sum()

Customer ID         0
Customer Name       0
City                0
Age                 0
Product Category    0
Sales               0
Order Status        0
dtype: int64

## Interpretation
The "df.isnull().sum()" function is for missing-value check which shows that all seven variables have zero missing values. Therefore, the dataset contains complete observations for Customer ID, Customer Name, City, Age, Product Category, Sales, and Order Status. No missing-value treatment is required at this stage.

In [13]:
df["Age"].unique()

array([65, 55, 19, 46, 45, 23, 20, 36, 60, 24, 40, 58, 70, 54, 34, 22, 18,
       66, 67, 41, 64, 31, 50, 69, 21, 25, 63, 38, 52, 33, 26, 49, 35, 62,
       44, 27, 53, 51, 48, 28, 43, 47, 39, 61, 57, 32, 59, 56, 68, 42, 37,
       30, 29])

### Investigation of Age Values

The unique values in the Age column are all numerical whole numbers, ranging from 18 to 70 years. No text or obviously invalid non-numerical values were identified in this column. Therefore, the Age variable does not require cleaning based on this inspection.

In [15]:
df["Sales"].unique()

array([ 61241.84,   7969.88,  49729.44, ..., 216110.54, 121369.5 ,
       155249.68], shape=(5000,))

### Investigation of Sales Values

The unique values in the Sales column are numerical values represented as floating-point numbers. No obvious text values or non-numerical entries were identified from the inspection. This is consistent with the `float64` data type reported during the initial dataset inspection. Therefore, no data-type correction is required for the Sales column based on this check.

In [16]:
df.duplicated().sum()

np.int64(0)

### Investigation of Duplicate Records

The duplicate-record check identified zero duplicated rows in the dataset. This indicates that no completely identical records were found during the initial data-quality assessment. Therefore, no records need to be removed on the basis of full-row duplication.

In [19]:
df["City"].value_counts()

City
Enugu            364
Abuja            360
Calabar          358
Warri            348
Uyo              346
Kaduna           338
Kano             337
Port Harcourt    336
Ilorin           331
Abeokuta         327
Benin City       325
Jos              321
Lagos            308
Owerri           303
Ibadan           298
Name: count, dtype: int64

### Distribution of Customers by City

The dataset contains records from 15 cities. Enugu has the highest number of customer records, with 364, followed by Abuja with 360 and Calabar with 358. Ibadan has the lowest number of records, with 298. The distribution indicates that customers are represented across all 15 cities, although the number of records varies between locations.

In [21]:
df["Product Category"].value_counts()

Product Category
Home Appliances    549
Office Supplies    524
Accessories        507
Books              505
Beauty             499
Clothing           497
Sports             484
Electronics        483
Furniture          478
Groceries          474
Name: count, dtype: int64

### Distribution of Product Categories

The dataset contains 10 product categories. Home Appliances is the most frequently represented category, with 549 records, while Groceries has the lowest frequency, with 474 records. The remaining categories have frequencies ranging from 478 to 524 records. This shows that the dataset contains records across a relatively broad range of product categories, with Home Appliances having the highest representation.

In [22]:
df["Order Status"].value_counts()

Order Status
Completed     3099
Pending        608
Processing     596
Cancelled      408
Returned       289
Name: count, dtype: int64

### Distribution of Order Status

The order-status distribution shows that 3,099 records are classified as completed, making Completed the most frequent order status in the dataset. There are 608 pending orders, 596 processing orders, 408 cancelled orders, and 289 returned orders. The frequencies sum to 5,000 records, which corresponds to the total number of records in the dataset. This indicates that all records have an identified order status.

In [23]:
df.isnull().any(axis=1).sum()

np.int64(0)

### Records Containing Missing Information

The analysis identified zero records containing missing information. This confirms that all 5,000 records have complete values across the seven variables examined. Therefore, no rows need to be removed or modified due to missing information.

In [25]:
df["Sales"].mean()

np.float64(124393.01743000002)

### Average Sales Value

The average sales value across the 5,000 records is approximately ₦124,393.02. This represents the mean sales value recorded in the dataset and provides a general indication of the typical sales value across the customer records.

## Part D: Basic Customer Analysis

This section presents the basic analysis of the e-commerce customer dataset. The analysis focuses on customer distribution by city, average sales value, product category frequency, completed orders, and records containing missing information.

In [26]:
df["City"].value_counts()

City
Enugu            364
Abuja            360
Calabar          358
Warri            348
Uyo              346
Kaduna           338
Kano             337
Port Harcourt    336
Ilorin           331
Abeokuta         327
Benin City       325
Jos              321
Lagos            308
Owerri           303
Ibadan           298
Name: count, dtype: int64

### Customer Distribution by City

The dataset contains customer records from 15 cities. Enugu has the highest number of records, with 364, while Ibadan has the lowest, with 298. The distribution shows that customer records are represented across all 15 cities, although the number of records varies by location.

In [27]:
df["Sales"].mean()

np.float64(124393.01743000002)

### Average Sales Value

The average sales value across the 5,000 customer records is approximately ₦124,393.02. This represents the mean sales value recorded in the dataset and provides an overall indication of the typical sales value across the customer records.

In [28]:
df["Product Category"].value_counts()

Product Category
Home Appliances    549
Office Supplies    524
Accessories        507
Books              505
Beauty             499
Clothing           497
Sports             484
Electronics        483
Furniture          478
Groceries          474
Name: count, dtype: int64

### Most Frequent Product Category

The product category with the highest frequency is Home Appliances, with 549 records. This is followed by Office Supplies with 524 records and Accessories with 507 records. Groceries has the lowest frequency, with 474 records. Therefore, Home Appliances is the most frequently represented product category in the dataset.

In [29]:
df["Order Status"].value_counts()

Order Status
Completed     3099
Pending        608
Processing     596
Cancelled      408
Returned       289
Name: count, dtype: int64

### Completed Orders

The analysis shows that 3,099 of the 5,000 records have a Completed order status. This makes Completed the most frequently recorded order status in the dataset, followed by Pending (608) and Processing (596).

In [30]:
df.isnull().any(axis=1).sum()

np.int64(0)

### Records Containing Missing Information

The analysis identified zero records containing missing information. This indicates that all 5,000 records have complete information across the seven variables examined. Therefore, no records need to be removed due to missing values.

## Challenge: Customer Value Classification

A new variable, `Customer_Value`, will be created to classify customer records according to their sales value. Records with sales of ₦100,000 or more will be classified as High Value, while records with sales below ₦100,000 will be classified as Standard.

In [31]:
df["Customer_Value"] = df["Sales"].apply(
    lambda x: "High Value" if x >+ 100000 else "Standard"
)

In [32]:
df.head()

,Customer ID,Customer Name,City,Age,Product Category,Sales,Order Status,Customer_Value
0,CUST00001,Joseph Obi,Lagos,65,Beauty,61241.84,Completed,Standard
1,CUST00002,Tolu Onyeka,Abuja,55,Home Appliances,7969.88,Completed,Standard
2,CUST00003,Fatima Ogunleye,Ilorin,19,Accessories,49729.44,Pending,Standard
3,CUST00004,Mary Akinyemi,Ibadan,46,Office Supplies,69565.72,Returned,Standard
4,CUST00005,Kelechi Abubakar,Owerri,45,Sports,69485.89,Completed,Standard


### Interpretation

The `Customer_Value` column was successfully added to the dataset. The classification rule correctly assigns records with sales values below ₦100,000 to the Standard category. The first five records all have sales values below the ₦100,000 threshold and are therefore classified as Standard.

In [35]:
df["Customer_Value"].value_counts()

Customer_Value
High Value    2983
Standard      2017
Name: count, dtype: int64

### Customer Value Classification

The customer-value classification shows that 2,983 records are classified as High Value, while 2,017 records are classified as Standard. This indicates that High Value records constitute the larger proportion of the dataset. Specifically, approximately 59.66% of the records meet or exceed the ₦100,000 sales threshold, while 40.34% have sales values below the threshold.